### PR TITLE
Ticket 1820 polygonal fourier

### LINKED ISSUE #1100: add tests for generic scattering calculator (Trac #1033)
The existing tests in sas_calculator/tests/utest_sas_gen.py only that the scattering calculator runs, not that it returns the correct value.

In particular, 1D, 1D averaging and 2D tests need to be performed.

Migrated from http://trac.sasview.org/ticket/1033
```json
{
    "status": "new",
    "changetime": "2017-11-07T18:24:02",
    "_ts": "2017-11-07 18:24:02.664118+00:00",
    "description": "The existing tests in sas_calculator/tests/utest_sas_gen.py only that the scattering calculator runs, not that it returns the correct value.\n\nIn particular, 1D, 1D averaging and 2D tests need to be performed.",
    "reporter": "pkienzle",
    "cc": "",
    "resolution": "",
    "workpackage": "SasView Bug Fixing",
    "time": "2017-11-07T18:24:02",
    "component": "SasView",
    "summary": "add tests for generic scattering calculator",
    "priority": "minor",
    "keywords": "",
    "milestone": "SasView 5.0.0",
    "owner": "",
    "type": "defect"
}
```

### LINKED ISSUE #1820: implementation of direct Fourier transform of polygonal areas and volumes
For the general scattering calculator allow the calculations of the Fourier transform over non-rectangular meshes, e.g. from finite element simulations.

### PR DESCRIPTION
**Pull request by rjbourne**
This pull request introduces the ability to use finite element meshes in the generic SANS calculator, in the form of legacy .vtk files. This closes #1820. The functionality is currently ready for testing - however the code still depends on pull request https://github.com/SasView/sasview/pull/1888, which is not expected to significantly change before merging - after which main will need to be re-merged into this branch to resolve any conflicts. Note that this significantly inflates the number of changes github thinks have been made.

The key changes in this PR are:

 - The introduction of a fourier transform method for finite element meshes - heavily based on this paper: https://arxiv.org/abs/2104.08309. This allows the transform of irregular meshes to be calculated.
 - The addition of a vtk loader - which is capable of loading in legacy .vtk files (as produced by e.g. netgen) (file spec. version 3.0 and below - currently only supporting the unstructured grid type.)
 - Updates to the documentation briefly describing the changes made. Larger changes to the documentation describing changes to the generic SANS calculator across multiple tickets are in progress on the branch  https://github.com/SasView/sasview/tree/scattering-calculator-docs
 - Unit tests for the generic scattering calculator. This adds tests both for the original functionality (fixes #1100) as well as the new functionality - along with a new test .vtk file representing a 2x2x2 cube formed from 5 tetrahedral elements.

### CODE REVIEW

**Comment by krzywon** [comment]
@bmaranville - This uses your FT implementation. Could you help out with a review?

**Comment by bmaranville** [comment]
Well how about that!  I'd be happy to look it over.

**Comment by dehoni** [comment]
ready for testing on Win

**Comment by dehoni** [comment]
ready for testing on Win

**Comment by rjbourne** [comment]
Adding the sinc function is definitely the nicer way to do that - I really should have spotted that while working on the special case for where the denominator was zero. I'll add that in and hopefully it should be a fairly smooth change.

I believe the eps offset is still required to handle cases where Qp itself is zero, as well as the denominator of the fraction, because of the factor of 1/Qp^2. As far as I can tell there isn't a nice way to handle this within the code as it currently stands because the cancellations that give a finite analytical answer happen between separate faces - e.g. on a cube the opposite pairs of faces both individually have diverging values, but added together the result is finite.

**Comment by wpotrzebowski** [comment]
@dehoni to contact Michael Adam to review

**Comment by butlerpd** [comment]
@dehoni expects an answer this week

**Comment by pkienzle** [comment]
It's annoying that there are two definitions of sinc: `sin(πx)/(πx)` and `sin(x)/x`. Numpy uses the first, and looking at the commit I see that you've discovered this.

File: src/sas/sascalc/calculator/sas_gen.py [hunk 1]
```diff
@@ -872,13 +1281,13 @@ def set_sldn(self, sld_n, non_zero_mag_only=True):
                               np.fabs(self.sld_my) +
                               np.fabs(self.sld_mz)).nonzero()
                 if len(is_nonzero[0]) > 0:
-                    self.sld_n = np.zeros_like(self.pos_x)
+                    self.sld_n = np.zeros((self.data_length))
```
[the remark this answers is not in the document]
Comment (dehoni) [reply src/sas/sascalc/calculator/sas_gen.py:1284]: Changed as self.data_length will be a simple scalar giving the length of pos_x

**Comment by smk78** [comment]
This PR appears to add a .vtk file for unit testing.

Please can we also add an example .vtk to the /src/sas/sasview/test folder, under /coordinate_data perhaps?

That file should then be described in /src/sas/sasview/test/media/testdata_help.rst.

And a line announcing .vtk should be added under 'Coordinate Formats' in /src/sas/qtgui/MainWindow/media/data_formats_help.rst

File: src/sas/qtgui/Calculators/GenericScatteringCalculator.py [hunk 2]
```diff
@@ -349,6 +350,13 @@ def verify_files_match(self):
         if self.verification_occurred:
             self.toggle_error_functionality()
             return
+        # ensure both files are point or element type- not a mixture
```
[the remark this answers is not in the document]
Comment (dehoni) [reply src/sas/qtgui/Calculators/GenericScatteringCalculator.py]: The commit [below](https://github.com/SasView/sasview/pull/1898/commits/273f2474261afd9409467a3210ebd9d61928a8ae) tries achieving to separate the data verification from the Gui code. The changes need further testing. The Unit test for the scattering calculator is about to be updated.

**Comment by dehoni** [comment]
> This PR appears to add a .vtk file for unit testing.
> 
> Please can we also add an example .vtk to the /src/sas/sasview/test folder, under /coordinate_data perhaps?
> 
> That file should then be described in /src/sas/sasview/test/media/testdata_help.rst.
> 
> And a line announcing .vtk should be added under 'Coordinate Formats' in /src/sas/qtgui/MainWindow/media/data_formats_help.rst

VTK appears as new coordinate format with a test file.

**Comment by llimeht** [comment]
Are `sphere_R=0_2025.vtk` and `five_tetrahedra_cube.vtk` in the preferred file format for future editing or is there another native file format that the software used to generate these files uses (such as VTK's preferred XML format)? If there's a preferred form for future modification, it's best that it is included too.

**Comment by dehoni** [comment]
> Are `sphere_R=0_2025.vtk` and `five_tetrahedra_cube.vtk` in the preferred file format for future editing or is there another native file format that the software used to generate these files uses (such as VTK's preferred XML format)? If there's a preferred form for future modification, it's best that it is included too.

For a range of typical programs like Mumax, OOMF, netgen, paraview etc there exist converters to the "simplest" VTK Legacy (.vtk extension) format. This is the reason why it is also the format to generate test data. 
The newer XML formats seem to support a few more features, however it is said to more complicated than the original VTK format, which might be a reason that for now I am not aware that it is widespread used.

**Comment by llimeht** [comment]
@dehoni, I can see that the VTK legacy format makes sense as the lowest common denominator format for interchange. 

Maybe an different example would help here: if this were a page of instructions in PDF format, I'd be asking if there was a Word/Libreoffice/TeX document that should be included in the repo. Noone would expect to edit that PDF at some time later, so the preferred format for modification (fix bugs, improve things, make derivative works etc) should be supplied too.

Returning to the vtk files, the comment "_converters_ to… VTK" seems to imply that there is a preferred native format for the data (that is not VTK) that was involved in creating it and would be used to make changes. If someone asked us to "change that sphere_R=0_2025.vtk a bit" would I start with that vtk file? Would you or @rjbourne start with that vtk file? Or is there some other file somewhere that makes a better starting point?

**Comment by dehoni** [comment]
I see the point. The mentioned micromagnetic simulation packages create omf files. A reader for this is already provided with examples. 
The VTK files were created independently with Netgen. The [documentation](https://github.com/SasView/sasview/blob/scattering-calculator-docs/src/sas/qtgui/Calculators/media/gsc_ex_magnetic_spheres.rst) contains an example with the script that allows to regenerate the test files.

**Pull request merged**